parser.js

```diff
@@ -1,185 +1,132 @@
 'use strict';
 
-module.exports = function parse(roomid, message) {
-	if (roomid.slice(0, 5) === 'view-') return parseChatPage(roomid, message);
+module.exports = function parse(roomid, messageType, parts) {
+	if (roomid.startsWith('view-')) return parseChatPage(roomid, messageType, parts);
 
-	const messages = message.split('\n');
-	let initalizing = false;
-	for (const m of messages) {
-		if (m.charAt(0) !== '|') {
-			return;
+	switch (messageType) {
+	case 'c:':
+		// we dont care about timestamps
+		parts.shift(); // eslint-disable-line no-fallthroughs
+	case 'chat':
+	case 'c':
+		parseChat(roomid, parts[0] === '~' ? '~' : parts[0].slice(1), parts.slice(1).join('|'));
+		break;
+	case 'raw':
+	case 'html':
+	case '': // raw message
+		break;
+	case 'join':
+	case 'j':
+	case 'J': {
+		const [auth, nick] = Tools.splitUser(parts[0]);
+		Rooms(roomid).userJoin(nick, auth);
+		break;
+	}
+	case 'leave':
+	case 'l':
+	case 'L':
+		Rooms(roomid).userLeave(parts[0].slice(1));
+		break;
+	case 'name':
+	case 'n':
+	case 'N': {
+		const [auth, newNick] = Tools.splitUser(parts[0]);
+		const oldNick = parts[1];
+		Users.renameUser(oldNick, auth, newNick);
+		Rooms(roomid).userRename(oldNick, auth, newNick);
+		break;
+	}
+	case 'uhtml':
+	case 'uhtmlchange':
+		break;
+	case 'pm':
+		parsePM(parts[0], parts.slice(2).join('|'));
+		break;
+	case 'init':
+		Rooms.addRoom(roomid, parts[0]);
+		break;
+	case 'title':
+		Rooms(roomid).setTitle(parts[0]);
+		break;
+	case 'users':
+		Rooms(roomid).updateUserlist(parts[0]);
+		break;
+	case 'deinit':
+		Rooms(roomid).destroy();
+		break;
+	case 'noinit':
+		if (parts[0] === 'joinfailed') {
+			const room = /You are banned from the room "(.*)"/.exec(parts[1]);
+			if (room) debug(`Join failed - Banned from room ${room[1]}`);
+		} else if (parts[1] === 'nonexistent') {
+			const room = /The room "(.*)" does not exist\./.exec(parts[1]);
+			if (room) debug(`Join failed - The room ${room[1]} does not exist or is modjoined`);
```

Comment:
You can just console.log things like this, I think even non-devs would want to see this kind of info.